program_intent_engagement/apps/api/v1/tests/test_views.py

```diff
@@ -0,0 +1,140 @@
+"""
+Tests for the program intent API views
+"""
+import json
+import uuid
+
+import ddt
+from django.urls import reverse
+
+from program_intent_engagement.apps.core.models import ProgramIntent
+from test_utils import ProgramIntentsAPITestCase
+
+
+@ddt.ddt
+class ProgramIntentsViewTests(ProgramIntentsAPITestCase):
+    """
+    Tests Program Intents View
+    """
+
+    def setUp(self):
+        super().setUp()
+
+        self.url = reverse('api:v1:program_intent-insert')
+
+        self.program_uuid = str(uuid.uuid4())
+        self.data = {
+            'program_uuid': self.program_uuid,
+            'reason': 'ALL_COURSES_COMPLETE',
+            'certainty': 'CERTAIN_YES',
+            'effective_timestamp': '2022-07-01 00:00:00',
+        }
+
+    def post_api(self, user, data):
+        """
+        Helper function to make a POST request to the API
+        """
+        data = json.dumps(data)
+        headers = self.build_jwt_headers(user)
+
+        return self.client.post(self.url, data, **headers, content_type="application/json")
+
+    def validate_response_code(self, user, data, expected_response_code):
+        """
+        Helper function to get API response and validate response code
+        """
+        response = self.post_api(user, data)
+        self.assertEqual(response.status_code, expected_response_code)
+        return response
+
+    def test_auth_failures(self):
+        """
+        Test that the endpoint correctly validates permissions.
+        """
+        # test with anonymous user without jwt headers
+        response = self.client.post(self.url)
+        self.assertEqual(response.status_code, 401)
+
+    def test_empty_program_intent(self):
+        """
+        Test that the endpoint returns 400 if input is empty
+        """
+        self.validate_response_code(self.user, {}, 400)
+
+    def test_valid_program_intent(self):
+        """
+        Test that the endpoint returns 200 if input is valid
+        """
+        self.validate_response_code(self.user, self.data, 200)
+        self.assertEqual(len(ProgramIntent.objects.all()), 1)
+        intent = ProgramIntent.objects.get(program_uuid=self.program_uuid,
+                                           reason='ALL_COURSES_COMPLETE',
+                                           certainty='CERTAIN_YES',
+                                           effective_timestamp='2022-07-01 00:00:00')
+        self.assertIsNotNone(intent)
+
+    # todo: due to change in model (unique together), should logic be different?
+    # todo: return 400 or 200?
+    def test_duplicate_program_intent(self):
```

Comment:
The serializer can't be seeing a duplicate because it has no brain, it's just turning json into an object with no state or outside reference. It's possible that what is going on here is the intent object has a bunch of fields from TimeStampedModel that you're not providing and no user will ever provide so you need to specify a more limited set of fields to serialize.